### PR TITLE
Fix restraints and restraint tests

### LINKED ISSUE #1089: Yank Binding affinity discrepancy between Yank 0.23.4 and Yank 0.20.1
Hi There, 

I was trying to calculate the binding affinity of Lysozyme bonded to Toluene. I was performing the calculation by using Yank 0.23.4. Unfortunately, the produced result is quite off compared the experimental value: Exp = ~ -5 kcal/mol vs Yank_0.23.4 = ~ -11 kcal/mol. I decided to run the same calculation by using the Yank version 0.20.1 which should be the last version before the sampling engine redesign and I’ve got a closer result Yank_0.20.1 = ~ -3.9 kcal/mol. Looking at the Yank reports, it is clear where this difference is probably coming from:

Yank 0.23.4

Free energy of binding  :   -18.524 +- 0.604 kT (-11.043 +- 0.360 kcal/mol)
DeltaG complex          :    26.034 +- 0.580 kT
DeltaG standard state correction:             -3.241 kT
DeltaG solvent          :     4.269 +- 0.169 kT
Enthalpy of binding     :    30.463 +- 1519.819 kT (18.161 +- 906.058 kcal/mol)   

Yank 0.20.1 

Free energy of binding  :    -6.623 +- 0.205 kT (-3.948 +- 0.122 kcal/mol)
DeltaG complex          :    25.620 +- 0.128 kT
DeltaG standard state correction:            -14.602 kT
DeltaG solvent          :     4.395 +- 0.160 kT
Enthalpy of binding     :    13.298 +- 18.044 kT (7.927 +- 10.757 kcal/mol)

The DeltaG standard state corrections are quite different while the two alchemically decoupling simulations are providing close energies. I have attached a link to a file with the runs for the two Yank versions tested (I’m sorry it is a quite large file).

Can you please look at the results and try to see where this issue is coming from? Thank you so much.

https://drive.google.com/file/d/1WGX1xtz_SnZLJXLaCKeuf3FgeW7xBiWb/view?usp=sharing

### PR DESCRIPTION
**Pull request by jchodera**
Fixes #1089

This PR fixes the incorrect standard state correction for Boresch-like restraints (#1089) and restraint test failures.

- [x] Get broken restraint unit tests running again
- [x] Update YAML to modern best practices
- [ ] See if we can get this to run on travis - **TOO SLOW**
- [x] Fix any detected issues

The issue was caused because `kT` was not converted to a unitless quantity before using it in in computing the integrand for quadrature in [this commit](https://github.com/choderalab/yank/commit/11570959a82effa9d2172b8616a493c1e18bddab).

More generally, we should probably *never* try to convert unit-bearing quantities to floating point quantities in a consistent unit system for computing integrands. We should always introduce a dimensionless value `x` that runs between 0 and 1 and multiply the integration range by this quantity. I haven't made that change here to minimize impact on other things.

I've also corrected #1091 and fixed a typo in the `mcmc` YAML docs that confused me while trying to update the YAML files.

### CODE REVIEW

**Comment by jchodera** [comment]
I switched to testing with SAMS, and I'm now running into this failure, where something seems to be missing from the NetCDF file:
```
2018-09-22 12:09:40,344: Computing covariance matrix...
2018-09-22 12:09:40,347: Deltaf_ij:
2018-09-22 12:09:40,347:    0.000   0.000
2018-09-22 12:09:40,347:    0.000   0.000
2018-09-22 12:09:40,347: dDeltaf_ij:
2018-09-22 12:09:40,347:    0.000   0.000
2018-09-22 12:09:40,347:    0.000   0.000
2018-09-22 12:09:40,347: Current Free Energy Estimate is 0.0 +- 0.0 kT
2018-09-22 12:09:40,349: Computing offline free energy estimate took    0.068s
2018-09-22 12:09:40,349: Iteration took 0.157s.
2018-09-22 12:09:40,349: Estimated completion in 0:01:09.564810, at Sat Sep 22 12:10:49 2018 (consuming total wall clock time 0:01:32.753080).
2018-09-22 12:09:40,445: Initial checkpoint file automatically chosen as /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpmkidbsxs/experiments/complex_checkpoint.nc
2018-09-22 12:09:40,468: analysis_particle_indices != on-file analysis_particle_indices!Using on file analysis indices of [ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26]
ERROR

======================================================================
ERROR: Test that the harmonic restraint simulated free energy equals the standard state correction
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/choderaj/github/choderalab/yank/yank/Yank/tests/test_restraints.py", line 209, in test_Boresch_free_energy
    general_restraint_run(options)
  File "/Users/choderaj/github/choderalab/yank/yank/Yank/tests/test_restraints.py", line 169, in general_restraint_run
    analyzer = multistate.ReplicaExchangeAnalyzer(ncfile)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/yank-0.23.5-py3.6-macosx-10.9-x86_64.egg/yank/multistate/multistateanalyzer.py", line 1151, in __init__
    super().__init__(*args, **kwargs)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/yank-0.23.5-py3.6-macosx-10.9-x86_64.egg/yank/multistate/multistateanalyzer.py", line 548, in __init__
    if not reporter.is_open():
  File "netCDF4/_netCDF4.pyx", line 2606, in netCDF4._netCDF4.Dataset.__getattr__
  File "netCDF4/_netCDF4.pyx", line 2552, in netCDF4._netCDF4.Dataset.getncattr
  File "netCDF4/_netCDF4.pyx", line 1188, in netCDF4._netCDF4._get_att
  File "netCDF4/_netCDF4.pyx", line 1638, in netCDF4._netCDF4._ensure_nc_success
AttributeError: NetCDF: Attribute not found
-------------------- >> begin captured logging << --------------------
```
Any ideas?

**Comment by jchodera** [comment]
Also, I seem to be stochastically running into this failure that I will debug separately:
```
2018-09-22 12:14:12,366: Iteration 434/1000
2018-09-22 12:14:12,366: Single node: executing <function SAMSSampler._mix_replicas at 0x1a3a370b70>
2018-09-22 12:14:12,366: Updating thermodynamic states using global-jump scheme...
ERROR

======================================================================
ERROR: Test that the harmonic restraint simulated free energy equals the standard state correction
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/choderaj/github/choderalab/yank/yank/Yank/tests/test_restraints.py", line 215, in test_Boresch_free_energy
    general_restraint_run(options)
  File "/Users/choderaj/github/choderalab/yank/yank/Yank/tests/test_restraints.py", line 170, in general_restraint_run
    yaml_builder.run_experiments()
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/yank-0.23.5-py3.6-macosx-10.9-x86_64.egg/yank/experiment.py", line 791, in run_experiments
    completed[exp_index] = self._run_experiment(exp)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/yank-0.23.5-py3.6-macosx-10.9-x86_64.egg/yank/experiment.py", line 3124, in _run_experiment
    built_experiment.run(n_iterations=switch_experiment_interval)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/yank-0.23.5-py3.6-macosx-10.9-x86_64.egg/yank/experiment.py", line 476, in run
    alchemical_phase.run(n_iterations=iterations_to_run)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/yank-0.23.5-py3.6-macosx-10.9-x86_64.egg/yank/yank.py", line 1209, in run
    self._sampler.run(n_iterations=n_iterations)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/yank-0.23.5-py3.6-macosx-10.9-x86_64.egg/yank/multistate/multistatesampler.py", line 674, in run
    self._mix_replicas()
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/yank-0.23.5-py3.6-macosx-10.9-x86_64.egg/yank/mpi.py", line 271, in _wrapper
    return run_single_node(rank, task, *args, **kwargs)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/yank-0.23.5-py3.6-macosx-10.9-x86_64.egg/yank/mpi.py", line 220, in run_single_node
    result = task(*args, **kwargs)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/yank-0.23.5-py3.6-macosx-10.9-x86_64.egg/yank/multistate/sams.py", line 379, in _mix_replicas
    self._global_jump(replicas_log_P_k)
  File "/Users/choderaj/miniconda/lib/python3.6/site-packages/yank-0.23.5-py3.6-macosx-10.9-x86_64.egg/yank/multistate/sams.py", line 459, in _global_jump
    new_state_index = np.random.choice(neighborhood, p=P_k)
  File "mtrand.pyx", line 1146, in mtrand.RandomState.choice
ValueError: probabilities do not sum to 1
-------------------- >> begin captured logging << --------------------
```

**Comment by jchodera** [comment]
@andrrizzi : This is ready for review.

File: Yank/restraints.py [hunk 1]
```diff
@@ -1738,6 +1738,10 @@ def determine_missing_parameters(self, thermodynamic_state, sampler_state, topog
             # Determine restraint parameters for these atoms.
             self._determine_restraint_parameters(sampler_state, topography)
 
+        # DEBUG
+        ssc = self.get_standard_state_correction(thermodynamic_state)
+        logger.debug('Standard state correction: {} kT'.format(ssc))
```
Comment (andrrizzi) [inline Yank/restraints.py]: Did you intend to get rid of this line before merging?

Comment (jchodera) [reply Yank/restraints.py]: No, I actually want that there, but I'll remove the `# DEBUG` comment.

**Comment by jchodera** [comment]
I'm running a complete set of nosetests on `lilac` now that may turn up more issues to fix before release, but local tests of the Boresch issue pass, so I'll merge this once travis tests complete.

**Pull request merged**